cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -3731,10 +3731,16 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
                     let cond_code = ctx.data(icmp).cond_code().unwrap();
                     CC::from_intcc(cond_code)
                 } else {
-                    // The input is a boolean value, compare it against zero.
+                    let sel_ty = ctx.input_ty(insn, 0);
                     let size = ctx.input_ty(insn, 0).bytes() as u8;
                     let test = put_input_in_reg(ctx, flag_input);
-                    ctx.emit(Inst::cmp_rmi_r(size, RegMemImm::imm(0), test));
+                    if sel_ty == types::B1 {
+                        // The input is a boolean value; test the LSB for nonzero.
+                        ctx.emit(Inst::test_rmi_r(size, RegMemImm::imm(1), test));
+                    } else {
+                        // The input is an integer; test the whole value for nonzero.
```

Comment:
Re: the nit -- yes, good point; it could be a wider-than-1-bit bool. I'm not sure what that would mean as an input to a select or brz/brnz, so I've just added an assert that we do not see another bool type in the else-branch.

Re: using `test` always -- yes, good point; actually `test rN, rN` is one byte shorter than `cmp rN, 0` so this is a code-size win too. (`rN, rN` and not `rN, 0` because this is an and-op; and 0xff...ff is much larger in emitted code than the R-R form).